ember-server/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala

```diff
@@ -241,13 +241,37 @@ private[server] object ServerHelpers {
                 requestVault)
             }
 
+            // TODO: fix the order of error handling?
             result.attempt.flatMap {
               case Right((req, resp, drain)) =>
-                send(socket)(Some(req), resp, idleTimeout, onWriteFailure) >>
-                  drain.map {
-                    case Some(nextBuffer) => Some(((req, resp), (nextBuffer, true)))
-                    case None => None
-                  }
+                // TODO: Should we pay this cost for every HTTP request?
```

Comment:
In Blaze, we opt into web socket support by wrapping another stage, but it's one more thing to trip over configuring for users.  I'm not sure the opt-in is worth it.

Reply:
I wonder if we could be a bit smarter here for keep-alive connections. If we know we're on one, the upgrade should never happen. I may do some performance tests later, but since this is on a 0.22 milestone I may just defer as a follow-up since opt-in is something we can add in later